Check all pydoc strings, fix problem with invalid rendered markdown

GNUSheep opened the pull request:
Fix issue: #169 

I think the best option is to use a little bit of markdown, but if u want to use pure pydoc strings, I can easily swap to ">>>" with keeping all the newlines.

Also sorry for not rendering docs, I don't use python often so I dunno how to make module usable in command line, and I don't wanna break your docs while using lazydocs API

Waiting for your respond
GNUSheep

francium commented:
Thanks for the PR.

You should checkout the [newly added contributing guide in the README](https://github.com/rustedpy/result#contributing). Specifically we'll want to run `make docs` and ensure the resulting `docs/result.md` looks correct. You can preview the markdown either on GitHub after committing and pushing. Or you can also setup VS Code and the "Markdown Preview Github Styling" extension and preview within VS Code.

And make sure to commit the updated `docs` folder after running `make docs`.

We also need to ensure it renders correctly in editors and IDEs. For this you can also use VS Code with the "Python" extension published by "ms-python". Then just hover over functions and methods and ensure the popup documentation is rendering as expected.

francium commented:
Also I'm not 100% sure about what syntax to use. The official Python docs are usually written in ReStructuredText I believe. However, LazyDocs may support Markdown docstring syntax.

We'll also need to ensure common editors like VS Code, PyCharm and Language Server's (Emacs, Neovim/Vim) render the docstrings correctly as well.

GNUSheep commented:
Okay, at first I want to apologize for not reading  [newly added contributing guide in the README](https://github.com/rustedpy/result#contributing). I did tests, removed all whitespaces, and updated docs.

Also I tested how does it look like in vscode.

While using a little bit of markdown it looks like this:

    ```python

    r: Result[int, str] = get_a_result()
    if is_ok(r):
        r   # r is of type Ok[int]
    elif is_err(r):
        r   # r is of type Err[str]

    ```

So the only way to remove that "````python" , it's to use ">>>". It's look way better in vscode, but it's looking kinda weird for lazy docs:

Usage: 

``` r: Result[int, str] = get_a_result()```

```if is_ok(r):```  

```    r   # r is of type Ok[int]```

``` elif is_err(r):```  

```     r   # r is of type Err[str]```

francium commented:
Take a look at the failing GitHub actions as well. flake8 is complaining about some code formatting issues. You can also run flake8 locally with `make lint`

https://github.com/rustedpy/result/actions/runs/7342229052/job/19992515725?pr=173

I'll try to review this PR in detail as soon as I'm able.

GNUSheep commented:
> Take a look at the failing GitHub actions as well. flake8 is complaining about some code formatting issues. You can also run flake8 locally with `make lint`
> 
> https://github.com/rustedpy/result/actions/runs/7342229052/job/19992515725?pr=173
> 
> I'll try to review this PR in detail as soon as I'm able.

Done that

francium commented on src/result/result.py:
````diff
@@ -505,11 +505,17 @@ def is_ok(result: Result[T, E]) -> TypeGuard[Ok[T]]:
+    elif is_err(r):
+        r   # r is of type Err[str]
+
+    ```
````
The codeblocks shouldn't have empty lines before and after them, they should be like this,

<pre><code>```python
r: Result[int, str] = get_a_result()
if is_ok(r):
    r   # r is of type Ok[int]
elif is_err(r):
    r   # r is of type Err[str]
```
</code></pre>

francium commented on docs/result.md:
```diff
@@ -118,28 +146,26 @@ do_async(
-         for z in get_sync_result_3()
-     )
+final_result: Result[float, int] = await do_async(  Ok(len(x) + int(y) + z)  for x in await get_async_result_1()  for y in await get_async_result_2()  for z in get_sync_result_3()  ) 
+
```
This rendered example doesn't look right. There's some issue with the docstring.

Tip, you can use this button to preview it in GitHub, in case you had not used it before
![image](https://github.com/rustedpy/result/assets/15352422/64649860-b70a-4bdc-ad3b-c55b1064f4cc)

francium commented:
I checked with PyCharm, it doesn't even support non-plaintext documents[1][2]. So we're going to ignore PyCharm and just make sure things render properly in VSCode and other Language Server based editor.

[1]: [reStructuredText Python code block in docstring does not render : PY-37743](https://youtrack.jetbrains.com/issue/PY-37743)
[2]: [Support "code" directive : PY-25658](https://youtrack.jetbrains.com/issue/PY-25658/Support-code-directive)

francium reviewed (approved):
Looks great. Thank you for the help @GNUSheep
